### PR TITLE
add support for openssl

### PR DESCRIPTION
**Pull request by poonai**
Signed-off-by: Tiger <rbalajis25@gmail.com>

### CODE REVIEW

**Comment by poonai** [comment]
I'll change the reference once this get merged https://github.com/libp2p/go-openssl/pull/6

**Comment by poonai** [comment]
Ive updated the go.mod @Stebalien could you check this?

**Comment by Stebalien** [comment]
cc @marten-seemann, could I ask you to review this?

File: crypto.go [hunk 1]
```diff
@@ -57,6 +46,25 @@ func NewIdentity(privKey ic.PrivKey) (*Identity, error) {
 	}, nil
 }
 
+func newIdentity(sk ic.PrivKey) (identity, error) {
+	return NewIdentity(sk)
+}
```
Comment (marten-seemann) [inline crypto.go:51]: Why is this additional constructor needed?

File: identity.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		peerID, err := peer.IDFromPublicKey(pubKey)
+		if err != nil {
+			peerID = peer.ID(fmt.Sprintf("(not determined: %s)", err.Error()))
+		}
+		return fmt.Errorf("peer IDs don't match: expected %s, got %s", remote, peerID)
+	}
+	return nil
+}
+
+func unmarshalExtenstionPublicKey(value []byte, certKeyPub []byte) (ic.PubKey, error) {
```
Comment (marten-seemann) [inline identity.go:48]: There's a typo in the function name.

File: identity.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	return &certificateConfig{
+		serialNumber:   sn,
+		extensionValue: value,
+		certKey:        certKey,
+	}, nil
+}
+
+// identity is used to secure connection.
+type identity interface {
```
Comment (marten-seemann) [inline identity.go:121]: Logically, this doesn't seem to belong to `Identity`. Also, it doesn't access any private members of `Identity`, so one could easily move it to a separate function.

File: openssl.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		return nil, nil, err
+	}
+
+	// Add the certificate.
+	if err := opensslCtx.UseCertificate(o.certificate); err != nil {
+		return nil, nil, err
+	}
+
+	// Enable two way tls.
+	opensslCtx.SetVerifyMode(openssl.VerifyPeer)
```
Comment (marten-seemann) [inline openssl.go:85]: I'm not familiar with the go-openssl API here. Why `VerifyPeer` and not `VerifyFailIfNoPeerCert`? What happens if the client doesn't present a certificate at all? And what is `VerifyClientOnce` anyway?

File: openssl.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+	opensslCtx.SetVerifyCallback(func(a bool, store *openssl.CertificateStoreCtx) bool {
+		cert := store.GetCurrentCert()
```
Comment (marten-seemann) [inline openssl.go:88]: `GetCurrentCert()` seems to return only a single certificate. What the peer presents though is a certificate **chain**. How do we get access to the chain (or at least make sure the whole chain is verified)?

**Review by marten-seemann — COMMENTED** [review]
After adding a number of detailed comments, I noticed that this PR is using `openssl.TLSv1_2`. Looking at [the code](https://github.com/libp2p/go-openssl/blob/d9b640282843752442a8f28637156e13b1abd3f4/ctx.go#L71-L82), it seems like go-openssl doesn't even support TLS 1.3 yet.

libp2p use of TLS is only specified for TLS 1.3, and negotiating any earlier TLS version is explicitly prohibited, see https://github.com/libp2p/specs/blob/master/tls/tls.md#handshake-protocol.

**Comment by marten-seemann** [comment]
Another thing (you don't have to implement this, @balajijinnah, and it doesn't need to be part of this PR), but before rolling this out into production we probably want to add an integration test that tests various handshake scenarios (both successful and unsuccessful) between an OpenSSL-enabled and an OpenSSL-disabled build.

This is not straightforward unfortunately, as the openssl flag has be set at build time, so we'll probably have to build two binaries and run them against each other.

What would be nice in this PR though is to split up the Travis build into two, so we can test in CI that at least the two handshake implementations interoperate among themselves.

**Comment by Geo25rey** [comment]
@balajijinnah Are you still working on this PR? If not, I'd be glad to continue with it.

**Comment by poonai** [comment]
@Geo25rey, please carry forward. I'm occupied now :(

File: openssl.go (hunk 4 above)
Comment (Geo25rey) [reply openssl.go:85]: See the [openssl man page]( https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html#NOTES). The go-openssl verify option bindings have a one-to-one correspondence to the C openssl bindings. I set it to both `VerifyPeer` and `VerifyFailIfNoPeerCert`.

File: identity.go (hunk 3 above)
Comment (Geo25rey) [reply identity.go:121]: Removed `type identity`

File: identity.go (hunk 2 above)
Comment (Geo25rey) [reply identity.go:48]: Fixed

File: crypto.go (hunk 1 above)
Comment (Geo25rey) [reply crypto.go:51]: It's not. Removed

File: openssl.go (hunk 5 above)
Comment (Geo25rey) [reply openssl.go:88]: The verify callback is run for every certificate in the chain.

Quote from [openssl man page](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html):
> The certificate chain is checked starting with the deepest nesting level (the root CA certificate) and worked upward to the peer's certificate. At each level signatures and issuer attributes are checked. Whenever a verification error is found, the error number is stored in x509_ctx and verify_callback is called with preverify_ok=0. By applying X509_CTX_store_* functions verify_callback can locate the certificate in question and perform additional steps (see EXAMPLES). If no error is found for a certificate, verify_callback is called with preverify_ok=1 before advancing to the next level.

**Pull request closed without merging**